### ISSUE TITLE
新的代码会导致OOM

### ISSUE THREAD

**Issue by mynewstart** [LianjiaTech/BELLE#393]
我使用没整合huggingface Trainer的code，使用deepspeed zero3训练llama-30b的模型，max_length=512, bz=4, gradient_accumulation_steps=8, 在A100 80G的卡上可以正常训练。但是我发现用现在的新代码，同样llama-30b的模型，max_length=512，同样使用zero3,但是bz=1都会报错OOM。 请问这主要是什么原因导致的？
### 旧code训练的commad：

OUTPUT="./output/actor-models/llama-30b-blend-data-test"
ZERO_STAGE=3

echo $OUTPUT
echo $ZERO_STAGE
data_output_path=./output/actor-models/data_files

deepspeed main.py \
   --sft_only_data_path xxx \
   --model_name_or_path decapoda-research/llama-30b-hf \
   --per_device_train_batch_size 4 \
   --per_device_eval_batch_size 1 \
   --max_seq_len 512 \
   --learning_rate 5e-6 \
   --weight_decay 0.0001 \
   --num_train_epochs 2  \
   --gradient_accumulation_steps 8 \
   --lr_scheduler_type cosine \
   --num_warmup_steps 0 \
   --seed 1234 \
   --gradient_checkpointing \
   --zero_stage $ZERO_STAGE \
   --deepspeed \
   --output_dir $OUTPUT \
   --data_output_path $data_output_path \

   ### 新code的训练脚本：
   torchrun --nproc_per_node 8 src/train.py \
    --model_name_or_path "decapoda-research/llama-30b-hf" \
    --llama \
    --deepspeed configs/deepspeed_config_stage3.json \
    --train_file xxx \
    --validation_file xxx \
    --per_device_train_batch_size 4 \
    --per_device_eval_batch_size 1 \
    --gradient_accumulation_steps 8 \
    --num_train_epochs 2 \
    --model_max_length 512 \
    --save_strategy "steps" \
    --save_total_limit 3 \
    --learning_rate 5e-6 \
    --weight_decay 0.0001 \
    --warmup_ratio 0.0 \
    --lr_scheduler_type "cosine" \
    --logging_steps 10 \
    --evaluation_strategy "steps" \
    --fp16 True \
    --seed 1234 \
    --output_dir xxx/ \
    使用的deepspeed config文件：
    `{
    "zero_optimization": {
        "stage": 3, 
        "offload_param": {
            "device": "none"
        }, 
        "offload_optimizer": {
            "device": "none"
        }, 
        "stage3_param_persistence_threshold": 1.000000e+04, 
        "stage3_max_live_parameters": 3.000000e+07, 
        "stage3_prefetch_bucket_size": 3.000000e+07, 
        "memory_efficient_linear": false
    },
    "fp16": {
        "enabled": true, 
        "loss_scale_window": 100
    },
    "optimizer": {
        "type": "AdamW",
        "params": {
          "lr": "auto",
          "betas": [
            0.9,
            0.999
          ],
          "eps": 1e-8,
          "weight_decay": "auto"
        }
    },
    "train_batch_size": "auto",
    "train_micro_batch_size_per_gpu": "auto",
    "wall_clock_breakdown": false,
    "gradient_accumulation_steps": "auto",
    "gradient_clipping": 1.0,
    "prescale_gradients": false, 
    "steps_per_print": 1000
}`

**Comment by xianghuisun** [comment]
> 我使用没整合huggingface Trainer的code，使用deepspeed zero3训练llama-30b的模型，max_length=512, bz=4, gradient_accumulation_steps=8, 在A100 80G的卡上可以正常训练。但是我发现用现在的新代码，同样llama-30b的模型，max_length=512，同样使用zero3,但是bz=1都会报错OOM。 请问这主要是什么原因导致的？
> 
> ### 旧code训练的commad：
> OUTPUT="./output/actor-models/llama-30b-blend-data-test" ZERO_STAGE=3
> 
> echo $OUTPUT echo $ZERO_STAGE data_output_path=./output/actor-models/data_files
> 
> deepspeed main.py --sft_only_data_path xxx --model_name_or_path decapoda-research/llama-30b-hf --per_device_train_batch_size 4 --per_device_eval_batch_size 1 --max_seq_len 512 --learning_rate 5e-6 --weight_decay 0.0001 --num_train_epochs 2 --gradient_accumulation_steps 8 --lr_scheduler_type cosine --num_warmup_steps 0 --seed 1234 --gradient_checkpointing --zero_stage $ZERO_STAGE --deepspeed --output_dir $OUTPUT --data_output_path $data_output_path \
> 
> ### 新code的训练脚本：
> torchrun --nproc_per_node 8 src/train.py --model_name_or_path "decapoda-research/llama-30b-hf" --llama --deepspeed configs/deepspeed_config_stage3.json --train_file xxx --validation_file xxx --per_device_train_batch_size 4 --per_device_eval_batch_size 1 --gradient_accumulation_steps 8 --num_train_epochs 2 --model_max_length 512 --save_strategy "steps" --save_total_limit 3 --learning_rate 5e-6 --weight_decay 0.0001 --warmup_ratio 0.0 --lr_scheduler_type "cosine" --logging_steps 10 --evaluation_strategy "steps" --fp16 True --seed 1234 --output_dir xxx/ 使用的deepspeed config文件： `{ "zero_optimization": { "stage": 3, "offload_param": { "device": "none" }, "offload_optimizer": { "device": "none" }, "stage3_param_persistence_threshold": 1.000000e+04, "stage3_max_live_parameters": 3.000000e+07, "stage3_prefetch_bucket_size": 3.000000e+07, "memory_efficient_linear": false }, "fp16": { "enabled": true, "loss_scale_window": 100 }, "optimizer": { "type": "AdamW", "params": { "lr": "auto", "betas": [ 0.9, 0.999 ], "eps": 1e-8, "weight_decay": "auto" } }, "train_batch_size": "auto", "train_micro_batch_size_per_gpu": "auto", "wall_clock_breakdown": false, "gradient_accumulation_steps": "auto", "gradient_clipping": 1.0, "prescale_gradients": false, "steps_per_print": 1000 }`

旧代码基于deepspeed-chat项目，可能是他们在deepspeed上又做了哪些优化吧

**Comment by mynewstart** [comment]
好的，请问你们也是有类似的体验吗？用之前deepspeed-chat的代码能训练更大的模型？
